mfai/torch/models/pangu.py

```diff
@@ -0,0 +1,888 @@
+        x = x.reshape(shape=(x.shape[0], Z*(H//2)*(W//2), -1))
+
+        # Call the layer normalization
+        x = self.norm(x)
+
+        # Decrease the size of the tokens to reduce computation cost
+        x = self.linear(x)
+        return x, torch.Size([x.shape[0], Z, (H//2), (W//2), x.shape[-1]])
+
+
+class UpSample(nn.Module):
+    def __init__(self, input_dim: int, output_dim: int) -> None:
+        """Up-sampling operation. The number of tokens is mutiplied by 4.
+
+        Args:
+            input_dim (int): input token size
+            output_dim (int): output token size
+        """
+        super().__init__()
+        # Linear layers without bias to increase channels of the data
+        self.linear1 = Linear(input_dim, output_dim*4, bias=False)
+
+        # Linear layers without bias to mix the data up
+        self.linear2 = Linear(output_dim, output_dim, bias=False)
+
+        # Normalization
+        self.norm = LayerNorm(output_dim)
+
+    def forward(self, x: Tensor, embedding_shape: torch.Size) -> Tensor:
+        assert x.shape[-1] % 4 == 0, "The token size must be divisible by 4, but is {}".format(
+            x.shape[-1])
+        # Z, H, W represent the desired output shape
+        h_d = embedding_shape[2] // 2 + embedding_shape[2] % 2
+        w_d = embedding_shape[3] // 2 + embedding_shape[3] % 2
+
+        # Call the linear functions to increase channels of the data
+        x = self.linear1(x)
+
+        # Reorganize x to increase the resolution: simply change the order and upsample from (8, 91, 180) to (8, 182, 360)
+        # Reshape x to facilitate upsampling.
+        x = x.reshape(shape=(x.shape[0], embedding_shape[1], h_d, w_d, 2, 2, -1))
+        # Change the order of x
+        x = x.permute((0, 1, 2, 4, 3, 5, 6))
+        # Reshape to get Tensor with a resolution of (8, 182, 360)
+        x = x.reshape(shape=(x.shape[0], embedding_shape[1], h_d*2, w_d*2, -1))
+
+        # Crop the output to the input shape of the network
+        x = x[:, :, :embedding_shape[2], :embedding_shape[3], :]
+
+        # Reshape x back
+        x = x.reshape(shape=(x.shape[0], -1, x.shape[-1]))
+
+        # Call the layer normalization
+        x = self.norm(x)
+
+        # Mixup normalized tensors
+        x = self.linear2(x)
+        return x
+
+
+class EarthSpecificLayer(nn.Module):
+    """Basic layer of our network, contains 2 or 6 blocks
+
+        Args:
+            depth (int): number of blocks
+            data_size (torch.Size): see EarthSpecificBlock
+            dim (int): see EarthSpecificBlock
+            drop_path_ratio_list (Tensor]): see EarthSpecificBlock
+            num_heads (int): see EarthSpecificBlock
+            window_size (Tuple[int, int, int], optional): see EarthSpecificBlock
+            dropout_rate (float, optional): see EarthSpecificBlock
+            checkpoint_activation (bool, optional): see EarthSpecificBlock
+            lam (bool, optional): see EarthSpecificBlock
+        """
```

Comment:
docstring indentation for consistency with codebase
```suggestion
    Args:
        depth (int): number of blocks
        data_size (torch.Size): see EarthSpecificBlock
        dim (int): see EarthSpecificBlock
        drop_path_ratio_list (Tensor]): see EarthSpecificBlock
        num_heads (int): see EarthSpecificBlock
        window_size (Tuple[int, int, int], optional): see EarthSpecificBlock
        dropout_rate (float, optional): see EarthSpecificBlock
        checkpoint_activation (bool, optional): see EarthSpecificBlock
        lam (bool, optional): see EarthSpecificBlock
    """
```